ipaclient/plugins/cert.py

```diff
@@ -32,12 +36,96 @@
 
 @register(override=True, no_fail=True)
 class cert_request(MethodOverride):
+    takes_options = (
+        Str(
+            'database?',
+            label=_('Path to NSS database'),
+            doc=_('Path to NSS database to use for private key'),
+        ),
+        Str(
+            'private_key?',
+            label=_('Path to private key file'),
+            doc=_('Path to PEM file containing a private key'),
+        ),
+        Flag(
+            'autogenerate',
+            label=_('Automatically generate CSR for request using LDAP data'),
+        )
+    )
+
     def get_args(self):
         for arg in super(cert_request, self).get_args():
             if arg.name == 'csr':
-                arg = arg.clone_retype(arg.name, File)
+                arg = arg.clone_retype(arg.name, File, required=False)
             yield arg
 
+    def forward(self, *keys, **options):
+        autogenerate = options.pop('autogenerate', False)
+        database = options.pop('database', None)
+        private_key = options.pop('private_key', None)
+
+        if autogenerate:
+            if database:
+                helper = u'certutil'
+                helper_args = ['-d', database]
+            elif private_key:
+                helper = u'openssl'
+                helper_args = [private_key]
+            else:
+                raise errors.InvocationError(
+                    format="One of 'database' or 'private_key' is required")
+
+            scriptfile = tempfile.NamedTemporaryFile(delete=False)
+            scriptfile.close()
+            csrfile = tempfile.NamedTemporaryFile(delete=False)
+            csrfile.close()
+            csrfilename = csrfile.name
+
+            # profile_id is optional for cert_request, but not for
+            # cert_get_requestdata, so pass the default explicitly when
+            # necessary
+            profile_id = options.get('profile_id')
+            if profile_id is None:
+                profile_id = self.get_default_of('profile_id')
+
+            requestdata = self.api.Command.cert_get_requestdata(
+                profile_id=profile_id,
+                principal=options.get('principal'),
+                out=unicode(scriptfile.name),
+                helper=helper)
+
+            helper_cmd = ['bash', '-e', scriptfile.name, csrfilename] + helper_args
+
+            try:
+                subprocess.check_call(helper_cmd)
+            except subprocess.CalledProcessError:
+                raise errors.CertificateOperationError(
+                    error=(
+                        _('Error running "%(cmd)s" to generate CSR') %
+                        {'cmd': ' '.join(helper_cmd)}))
```

Comment:
It would be nice to at least log the original error, or use it as part of the new error message. Ditto for the `IOError` below.